File: app/app/src/main/java/com/example/hakonsreader/api/responses/JsonResponse.java
package com.example.hakonsreader.api.responses;

import com.example.hakonsreader.api.model.RedditListing;
import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * Handles responses in the form of:
 * <pre>
 * "json": {
 *     "errors": [],
 *     "data:" {
 *
 *     }
 * }
 * </pre>
 */
public class JsonResponse {

    @SerializedName("json")
    private ListingResponse response;

    /**
     * @return The list of more response
     */
    public List<? extends RedditListing> getListings() {
        return response.getListings();
    }

    public boolean hasErrors() {
        return response.hasErrors();
    }

    public List<List<String>> errors() {
        return response.getErrors();
    }
}

File: app/app/src/main/java/com/example/hakonsreader/api/responses/JsonResponse.kt
package com.example.hakonsreader.api.responses

import com.google.gson.annotations.SerializedName

/**
 * Response class for responses that wrap a [ListingResponse] with a "json: {}" object
 *
 * The responses will look like:
 *
 * "json": {
 *     "errors": [],
 *     "data:" {
 *
 *     }
 * }
 *
 */
class JsonResponse<T> {

    @SerializedName("json")
    private var response: ListingResponse<T>? = null

    /**
     * @return The list of more response
     */
    fun getListings(): List<T?>? {
        return response?.getListings()
    }

    fun hasErrors(): Boolean {
        return response?.hasErrors() == true
    }

    fun errors(): List<List<String?>?>? {
        return response?.getErrors()
    }
}